crates/solidity/inputs/language/src/definition.rs

```diff
@@ -0,0 +1,3854 @@
+                                ),
+                                Variant(
+                                    name = Library,
+                                    fields =
+                                        (definition = Required(NonTerminal(LibraryDefinition)))
+                                ),
+                                Variant(
+                                    name = Struct,
+                                    enabled_in = "0.6.0",
+                                    fields = (definition = Required(NonTerminal(StructDefinition)))
+                                ),
+                                Variant(
+                                    name = Enum,
+                                    enabled_in = "0.6.0",
+                                    fields = (definition = Required(NonTerminal(EnumDefinition)))
+                                ),
+                                Variant(
+                                    name = Function,
+                                    enabled_in = "0.7.1",
+                                    fields =
+                                        (definition = Required(NonTerminal(FunctionDefinition)))
+                                ),
+                                Variant(
+                                    name = Constant,
+                                    enabled_in = "0.7.4",
+                                    fields =
+                                        (definition = Required(NonTerminal(ConstantDefinition)))
+                                ),
+                                Variant(
+                                    name = Error,
+                                    enabled_in = "0.8.4",
+                                    fields = (definition = Required(NonTerminal(ErrorDefinition)))
+                                ),
+                                Variant(
+                                    name = UserDefinedValueType,
+                                    enabled_in = "0.8.8",
+                                    fields = (definition =
+                                        Required(NonTerminal(UserDefinedValueTypeDefinition)))
+                                ),
+                                Variant(
+                                    name = Using,
+                                    enabled_in = "0.8.13",
+                                    fields = (directive = Required(NonTerminal(UsingDirective)))
+                                )
+                            ]
+                        )
+                    ]
+                ),
+                Topic(
+                    title = "Pragma Directives",
+                    lexical_context = Pragma,
+                    items = [
+                        Struct(
+                            name = PragmaDirective,
+                            fields = (
+                                pragma_keyword = Required(Terminal([PragmaKeyword])),
```

Comment:
This seems overly verbose - can `Required~ be the default, and can single element arrays be replaced by list the element?

Reply:
Should be able to have these fine controls once we move to serde, using things like `#[serde(flatten)]` and `#[serde(default)]`, but for now, I suggest keeping the hand-written parser simple/straight-forward. It is expensive to add more capabilities for it only to throw it away later.